ext/test/zpages/tracez_data_aggregator_tests.cc

```diff
@@ -0,0 +1,531 @@
+  ASSERT_EQ(data.size(),span_name_to_duration.size());
+  
+  for(auto& span: span_name_to_duration)
+  {
+    ASSERT_TRUE(data.find(span.first) != data.end());
+    auto& aggregated_data = data.at(span.first);
+    
+    //Check if latency samples are in correct boundaries
+    for(LatencyBoundaryName boundary = LatencyBoundaryName::k0MicroTo10Micro; boundary != LatencyBoundaryName::k100SecondToMax; ++boundary)
+    {
+        ASSERT_EQ(aggregated_data->span_count_per_latency_bucket[boundary], span.second[boundary].size());
+        ASSERT_EQ(aggregated_data->latency_sample_spans[boundary].size(),span.second[boundary].size());
+        
+        auto latency_sample = aggregated_data->latency_sample_spans[boundary].begin();
+        for(unsigned int idx = 0 ; idx < span.second[boundary].size(); idx++)
+        {
+          ASSERT_EQ(span.second[boundary][idx],latency_sample->get()->GetDuration());
+          latency_sample = std::next(latency_sample);
+        }
+    }
+    
+    //Check if nothing else is affected
+    ASSERT_TRUE(aggregated_data->error_sample_spans.empty());
+    ASSERT_EQ(aggregated_data->num_running_spans,0);
+    ASSERT_EQ(aggregated_data->num_error_spans,0);
+  }
+}
+
+
+/** 
+ * This test checks to see if the aggregated data is updated correctly when there are multiple error spans.
+ * It checks both the count of error spans and the error samples
+ */ 
+TEST(TraceZDataAggregator, MultipleErrorSpans)
+{
+  std::shared_ptr<TracezSpanProcessor> processor(new TracezSpanProcessor());
+  auto tracer = std::shared_ptr<opentelemetry::trace::Tracer>(new Tracer(processor));
+  auto tracez_data_aggregator (new TracezDataAggregator(processor));
+  
+  //Container to store the span names --> error messges for the span name
+  std::unordered_map<std::string, std::vector<std::string>> span_name_to_error({
+    {span_name1, {"span 1 error"}},
+    {span_name2, {"span 2 error 1", "span 2 error 2"}},
+    {span_name3, {"span 3 error 1", "span 3 error 2", "span 3 error 3", "span 3 error 4", "span 3 error 5"}}
+  });
+  
+  //Start spans with the error messages based on the map
+  for(auto& span_error: span_name_to_error)
+  {
+    for(auto error_desc: span_error.second)tracer->StartSpan(span_error.first)->SetStatus(opentelemetry::trace::CanonicalCode::CANCELLED,error_desc);
+  }
+
+  const std::map<std::string, std::unique_ptr<AggregatedSpanData>>& data = tracez_data_aggregator->GetAggregatedData();
+  ASSERT_EQ(data.size(),span_name_to_error.size());
+  
+  //Check if error spans were updated correctly for the different span names
+  for(auto& span_error: span_name_to_error)
+  {
+    // First try to find the span name in aggregation, then check the count of the error spans and then check values
+    ASSERT_TRUE(data.find(span_error.first) != data.end());
+    
+    auto& aggregated_data = data.at(span_error.first);
+    
+    ASSERT_EQ(aggregated_data->num_error_spans, (int)span_error.second.size()); 
```

Comment:
Just curious, does it fail in compilation without the casting?

Reply:
No but I will be comparing unsigned int with a int.